doc/adr007-flattening.md

```diff
@@ -0,0 +1,255 @@
+
+All of this tinkering on names and the addition of a matching import statement makes it so the module outputted by the instance flattener can be name resolved. The name resolution of this module will produce a lookup table where the names referring to instances point to the new modules (produced for each instance), which are no longer instances. Therefore, after instance flattening, we have instance-less modules and their corresponding lookup table.
+
+#### Example
+
+Take this module as example, where `A` is instanced twice with the same name:
+```bluespec
+module A {
+  const N: int // id 1
+  val a = N // id 2
+}
+
+module B {
+  import A(N=1) as A1
+  val b = A1::a // id 3
+}
+
+module C {
+  import A(N=0) as A1
+  val C = A1::a // id 4
+}
+```
+
+The instance flattener will create two new modules: `B::A1` and `C::A1`, with new ids for the definitions inside it.
+```bluespec
+module A {
+  const N: int // id 1
+  val a = N // id 2
+}
+
+module B::A1 {
+  pure val B::A1::N = 1 // id 5
+  pure val B::A1::a = B::A1::N // id 6
+}
+
+module C::A1 {
+  pure val C::A1::N = 0 // id 7
+  pure val C::A1::a = C::A1::N // id 8
+}
+
+module B {
+  import B::A1.*
+  val b = B::A1::a // id 3
+}
+
+module C {
+  import C::A1.**
+  val c = C::A1::a // id 4
+}
+```
+
+Lookup table before instance flattening:
+```mermaid
+graph BT
+   1["(1) const N: int"]
+   2["(2) val a = N"] --> 1
+   3["(3) val b = A1::a"] --> 2
+   4["(4) val c = A1::a"] --> 2
+```
+
+After instance flattening, there are two independent sub-trees, since both instances are completely independent of each other:
+```mermaid
+graph BT
+  1["(1) const N: int"]
+  2["(2) val a = N"] --> 1
+  3["(3) val b = B::A1::a"] --> 6
+  4["(4) val c = C::A1::a"] --> 8
+  5["(5) pure val B::A1::N = 1"]
+  6["(6) pure val B::A1::a = B::A1::N"] --> 5
+  7["(7) pure val C::A1::N = 0"]
+  8["(8) pure val C::A1::a = C::A1::N"] --> 7
+```
+
+##### Override expressions
+
+In an instance statement like `import A(N=e).*`, `e` might depend on expressions defined in the current module. In this case, we need to ensure that the dependencies of `e` are copied to the new module created for the instance, otherwise `pure val myModule::A::N = e` won't resolve. To accomplish this, we use the actual `Flattener` (described below) to inspect only the expressions in the overrides and find any definitions they might depend on.
```

Comment:
Having to copy values "both ways" -- both into the importing module and into the instantiated module -- sounds a bit tricky and error prone, esp. considering that the instantiated module will get copied back into the importing module, meaning well be duplicating the definition of `e`.  Do I understand correctly, that given

```quint
module A {
  const a : int
}

module B {
   val e = 42
   import A(a = e).*
}
``` 

the proposed approach for "instance flattening" will end up giving us the intermediate result

```quint
module A {
  const a : int
}

module B::A {
  pure val B::A::e 
  pure val B::A::a : int = B::A::e
}

module B {
   pure val e = 42
   import B::A.*
}
``` 

and then this will be finally turned on the flattening pass into

```quint
module B {
   pure val e = 42
   pure val B::A::e 
   pure val B::A::a : int = B::A::e
}
```

Is that right?

Reply:
Yes, I believe that this is the case.

(I'm assuming you meant `pure val B::A::e = 42` instead of only `pure val B::A::e` since that's not even a valid definition)